### PR TITLE
Bind XML with no grammar constraints to XSD / DTD

### LINKED ISSUE #151: Bind XML with no grammar constraints to XSD / DTD
IntelliJ IDEA has this command to automatically generate an XSD schema from an XML file:

generate-schema

Such command would take at least 2 parameters:
- source file URI
- target XSD file URI (optional, would default to source name without extension + ".xsd")

I'd give that issue a very low priority, until other people vote for it.

### PR DESCRIPTION
**Pull request by angelozerr**
Bind XML with no grammar constraints to XSD / DTD

Fixes #151

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
To play with this PR, 

create a XML without grammar binding like:

```xml
<?xml version="1.0"?>
<catalog> 
	<book id="bk101"> 
		<author>Gambardella, Matthew</author>
		<title>XML Developer's Guide</title>
		<genre>Computer</genre>
		<price>44.95</price>
		<publish_date>2000-10-01</publish_date>
		<description>An in-depth look at creating applications 
      with XML.</description>
	</book>
	<book id="bk102">
		<author>Ralls, Kim</author>
		<title>Midnight Rain</title>
		<genre>Fantasy</genre>
		<price>5.95</price>
		<publish_date>2000-12-16</publish_date>
		<description>A former architect battles corporate zombies, 
      an evil sorceress, and her own childhood to become queen 
      of the world.</description>
	</book>
	<book id="bk103">
		<author>Corets, Eva</author>
		<title>Maeve Ascendant</title>
		<genre>Fantasy</genre>
		<price>5.95</price>
		<publish_date>2000-11-17</publish_date>
		<description>After the collapse of a nanotechnology 
      society in England, the young survivors lay the 
      foundation for a new society.</description>
	</book>
	<book id="bk104">
		<author>Corets, Eva</author>
		<title>Oberon's Legacy</title>
		<genre>Fantasy</genre>
		<price>5.95</price>
		<publish_date>2001-03-10</publish_date>
		<description>In post-apocalypse England, the mysterious 
      agent known only as Oberon helps to create a new life 
      for the inhabitants of London. Sequel to Maeve 
      Ascendant.</description>
	</book>
	<book id="bk105">
		<author>Corets, Eva</author>
		<title>The Sundered Grail</title>
		<genre>Fantasy</genre>
		<price>5.95</price>
		<publish_date>2001-09-10</publish_date>
		<description>The two daughters of Maeve, half-sisters, 
      battle one another for control of England. Sequel to 
      Oberon's Legacy.</description>
	</book>
	<book id="bk106">
		<author>Randall, Cynthia</author>
		<title>Lover Birds</title>
		<genre>Romance</genre>
		<price>4.95</price>
		<publish_date>2000-09-02</publish_date>
		<description>When Carla meets Paul at an ornithology 
      conference, tempers fly as feathers get ruffled.</description>
	</book>
	<book id="bk107">
		<author>Thurman, Paula</author>
		<title>Splish Splash</title>
		<genre>Romance</genre>
		<price>4.95</price>
		<publish_date>2000-11-02</publish_date>
		<description>A deep sea diver finds true love twenty 
      thousand leagues beneath the sea.</description>
	</book>
	<book id="bk108">
		<author>Knorr, Stefan</author>
		<title>Creepy Crawlies</title>
		<genre>Horror</genre>
		<price>4.95</price>
		<publish_date>2000-12-06</publish_date>
		<description>An anthology of horror stories about roaches,
      centipedes, scorpions  and other insects.</description>
	</book>
	<book id="bk109">
		<author>Kress, Peter</author>
		<title>Paradox Lost</title>
		<genre>Science Fiction</genre>
		<price>6.95</price>
		<publish_date>2000-11-02</publish_date>
		<description>After an inadvertant trip through a Heisenberg
      Uncertainty Device, James Salway discovers the problems 
      of being quantum.</description>
	</book>
	<book id="bk110">
		<author>O'Brien, Tim</author>
		<title>Microsoft .NET: The Programming Bible</title>
		<genre>Computer</genre>
		<price>36.95</price>
		<publish_date>2000-12-09</publish_date>
		<description>Microsoft's .NET initiative is explored in 
      detail in this deep programmer's reference.</description>
	</book>
	<book id="bk111">
		<author>O'Brien, Tim</author>
		<title>MSXML3: A Comprehensive Guide</title>
		<genre>Computer</genre>
		<price>36.95</price>
		<publish_date>2000-12-01</publish_date>
		<description>The Microsoft MSXML3 parser is covered in 
      detail, with attention to XML DOM interfaces, XSLT processing, 
      SAX and more.</description>
	</book>
	<book id="bk112">
		<author>Galos, Mike</author>
		<title>Visual Studio 7: A Comprehensive Guide</title>
		<genre>Computer</genre>
		<price>49.95</price>
		<publish_date>2001-04-16</publish_date>
		<description>Microsoft Visual Studio 7 is explored in depth,
      looking at how Visual Basic, Visual C++, C#, and ASP+ are 
      integrated into a comprehensive development 
      environment.</description>
	</book>
</catalog>
```

and generate a XSD , DTD bindings with code action. see the following demo:

NoGrammarConstraintsDemo

**Comment by angelozerr** [comment]
@xorye @datho7561 please review my PR. Thanks!

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/CodeActionFactory.java [hunk 1]
```diff
@@ -67,35 +67,39 @@ public static CodeAction insert(String title, Position position, String insertTe
...
 	}
 
+	public static TextDocumentEdit insertEdit(String insertText, Position position, TextDocumentItem document) {
+		TextEdit edit = new TextEdit(new Range(position, position), insertText);
+		VersionedTextDocumentIdentifier versionedTextDocumentIdentifier = new VersionedTextDocumentIdentifier(
+				document.getUri(), document.getVersion());
+
+		TextDocumentEdit textDocumentEdit = new TextDocumentEdit(versionedTextDocumentIdentifier,
+				Collections.singletonList(edit));
+		return textDocumentEdit;
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/CodeActionFactory.java:84]: `return new TextDocumentEdit(versionedTextDocumentIdentifier, Collections.singletonList(edit));`

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+			String schemaFileName = getFileName(schemaURI);
+			String schemaTemplate = generator.generate(document, sharedSettings, new XMLSchemaGeneratorSettings());
+
+			// xsi:noNamespaceSchemaLocation
+			// Create code action to create the XSD file with the generated XSD content
+			String insertText = " xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\""
+					+ document.getTextDocument().lineDelimiter(0);
+			insertText += " xsi:noNamespaceSchemaLocation=\"" + schemaFileName + "\"";
+			CodeAction noNamespaceSchemaLocationAction = createGrammarFileAndBoundIt(
+					"Bind to a XML Schema with xsi:noNamespaceSchemaLocation", schemaURI, schemaTemplate, insertText,
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java:53]: Bind to an XML

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+					afterTagOffset, document, diagnostic);
+			codeActions.add(noNamespaceSchemaLocationAction);
+
+			// xml-model
+			XMLBuilder xsdWithXmlModel = new XMLBuilder(sharedSettings, null, delimiter);
+			xsdWithXmlModel.startPrologOrPI("xml-model");
+			xsdWithXmlModel.addSingleAttribute("href", schemaFileName, true);
+			xsdWithXmlModel.endPrologOrPI();
+			xsdWithXmlModel.linefeed();
+			CodeAction xsdWithXmlModelAction = createGrammarFileAndBoundIt("Bind to a XSD with xml-model", schemaURI,
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java:63]: ... "Bind to an XSD ...

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/CodeActionFactory.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/commons/CodeActionFactory.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java (hunk 3 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java (hunk 2 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java]: I changed to XSD to be conistent., fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+		} catch (BadLocationException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+
+	private static String getGrammarURI(String documentURI, String fileExtension) {
+		int index = documentURI.lastIndexOf('.');
+		String grammarURI = documentURI.substring(0, index);
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java:118]: Could we check for `index == -1`?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		grammarURI += "." + fileExtension;
+		return grammarURI;
+	}
+
+	private static String getFileName(String schemaURI) {
+		return new File(schemaURI).getName();
+	}
+
+	private static CodeAction createGrammarFileAndBoundIt(String title, String grammarURI, String grammarContent,
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java:127]: createGrammarFileAndBindIt

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+			String insertText, int insertOffset, DOMDocument document, Diagnostic diagnostic)
+			throws BadLocationException {
+		Position position = document.positionAt(insertOffset);
+		TextDocumentEdit insertEdit = CodeActionFactory.insertEdit(insertText, position, document.getTextDocument());
+		return createGrammarFileAndBoundIt(title, grammarURI, grammarContent, insertEdit, diagnostic);
+	}
+
+	private static CodeAction createGrammarFileAndBoundIt(String title, String grammarURI, String grammarContent,
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java:135]: createGrammarFileAndBindIt

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java (hunk 5 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java (hunk 6 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java (hunk 4 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/NoGrammarConstraintsCodeAction.java]: fixed and I have improved this method to return an unique file name. I have improved too code action title to show the generated grammar file name.

**Pull request merged**